packages/tests/server/regression/issue-5017-inference-middleware.test.ts

```diff
@@ -0,0 +1,52 @@
+import { inferRouterInputs, inferRouterOutputs, initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+const t = initTRPC.create();
+const appRouter = t.router({
+  str: t.procedure.input(z.string()).query(({ input }) => input),
+  strWithMiddleware: t.procedure
+    // ^?
+    .input(z.string())
+    .use((opts) => opts.next())
+    .query(({ input }) => input),
+
+  objWithMiddleware: t.procedure
+    .input(
+      z.object({
+        foo: z.string(),
+      }),
+    )
+    .use((opts) => opts.next())
+    .query(({ input }) => input),
+});
+type AppRouter = typeof appRouter;
+
+describe('inferRouterInputs', () => {
+  type AppRouterInputs = inferRouterInputs<AppRouter>;
+  type AppRouterOutputs = inferRouterOutputs<AppRouter>;
+
+  test('string', async () => {
+    {
+      type Input = AppRouterInputs['str'];
+      type Output = AppRouterOutputs['str'];
+      expectTypeOf<Input>().toBeString();
+      expectTypeOf<Output>().toBeString();
+    }
+    {
+      type Input = AppRouterInputs['strWithMiddleware'];
+      //   ^?
+      type Output = AppRouterOutputs['strWithMiddleware'];
+      expectTypeOf<Input>().toBeString();
+      expectTypeOf<Output>().toBeString();
```

Comment:
This is the one that's failing

Reply:
I think the proper solution would be a sane implementation of `Overwrite` that only does things you'd expected. Example pseudo:

```
Given A and B,

- If they are both objects, overwrite A's keys with B
- If only A is object, return A
- If only B is object, return B
- If A extends any, return A
- If B extends any, return B
- Otherwise return never
```

But are there any other practical usecases for Overwrite?

Reply:
I think it makes more sense for an intuitive meaning of "Overwrite" so that key-wise overwrite happens when both are objects, and otherwise `TWith` just replaces `TType` entirely.